### PR TITLE
Add Cosmopedia blog post

### PR DESCRIPTION
**Pull request by loubnabnl**

### CODE REVIEW

File: cosmopedia.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+```
+
+You can even track the generations with `wandb` to monitor the throughput and number of generated tokens.
+<p align="center">
+ <img src="assets/cosmopedia/wandb.png" alt="text-clustering" style="width: 60%; height: auto;"><br>
+ <em>Figure 10. Wandb plots for an llm-swarm run.</em>
+</p>
+
+**Note:**
+We used HuggingChat for the initial iterations on the prompts. Then, we ran each prompt on a few hundred samples using `llm-swarm` to spot unusual patterns. For instance, the model used very similar introductory phrases for textbooks and frequently began stories with the same phrases, like "Once upon a time" and "The sun hung low in the sky". Explicitly asking the model to avoid these introductory statements and to be creative fixed the issue; they were still used but less frequently.
```
Comment (anton-l) [inline cosmopedia.md:158]:
```suggestion
We used HuggingChat for the initial iterations on the prompts. Then, we generated a few hundred samples for each prompt using `llm-swarm` to spot unusual patterns. For instance, the model used very similar introductory phrases for textbooks and frequently began stories with the same phrases, like "Once upon a time" and "The sun hung low in the sky". Explicitly asking the model to avoid these introductory statements and to be creative fixed the issue; they were still used but less frequently.
```

File: cosmopedia.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+#### Benchmark decontamination
+
+Given How we generate synthetic data, there is a possibility of benchmark contamination within the seed samples or the model's training data. To address this, we implement a decontamination pipeline to ensure our dataset is free of any samples from the test benchmarks.
```
Comment (anton-l) [inline cosmopedia.md:162]:
```suggestion
Given that we generate synthetic data, there is a possibility of benchmark contamination within the seed samples or the model's training data. To address this, we implement a decontamination pipeline to ensure our dataset is free of any samples from the test benchmarks.
```

File: cosmopedia.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+The model performs better than TinyLlama 1.1B on ARC-easy, ARC-challenge, OpenBookQA, and MMLU and is comparable to Qwen-1.5-1B on ARC-challenge and OpenBookQA. However, we notice some performance gaps compared to Phi-1.5, suggesting a better synthetic generation quality, which can be related to the LLM used for generation, topic coverage, or prompts.
+
+<p align="center">
+ <img src="assets/cosmopedia/evals.png" alt="evaluation results" style="width: 60%; height: auto;"><br>
+ <em>Figure 10. Evaluation results of Cosmo-1B.</em>
+</p>
+
+## Conclusion & next steps
+
+In this blog, we outlined  has our approach for creating Cosmopedia, a large synthetic dataset designed for pre-training models, with the goal of replicating the Phi datasets. We highlighted the significance of meticulously crafting prompts to cover a wide range of topics, ensuring the generation of diverse content. Additionally, we have shared and open-sourced our technical stack, which allows for scaling the generation process across hundreds of GPUs.
```
Comment (anton-l) [inline cosmopedia.md:195]:
```suggestion
In this blog, we outlined our approach for creating Cosmopedia, a large synthetic dataset designed for pre-training models, with the goal of replicating the Phi datasets. We highlighted the significance of meticulously crafting prompts to cover a wide range of topics, ensuring the generation of diverse content. Additionally, we have shared and open-sourced our technical stack, which allows for scaling the generation process across hundreds of GPUs.
```

**Review by anton-l — APPROVED** [review]

File: cosmopedia.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
+---
+title: "Cosmopedia: how to create large-scale synthetic data for pre-training"
```
Comment (davanstrien) [inline cosmopedia.md:2]:
```suggestion
title: "Cosmopedia: how to create large-scale synthetic data for pre-training Large Language Models"
```
Maybe it makes the scope a bit clearer but I don't have a strong preference for the current title.

File: cosmopedia.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+<p align="center">
+ <img src="https://huggingface.co/datasets/HuggingFaceTB/images/resolve/main/cosmopedia/data.png" alt="number of datasets with synthetic tag" style="width: 90%; height: auto;"><br>
+<em>Figure 1. Datasets on Hugging Face hub with the tag synthetic.</em>
+</p>
+
+However, this is not another blog post on generating synthetic instruction-tuning datasets, a subject that the community is already extensively exploring. Our focus is on scaling from a **few thousand** to **millions** of samples that can be used for **pre-training LLMs from scratch**, presenting a unique set of challenges. This blog post is a follow-up to our first introductory blog to the world of synthetic datasets [TODO: add a link to Daniel’s blog].
```
Comment (davanstrien) [inline cosmopedia.md:20]:
```suggestion
However, this is not another blog post on generating synthetic instruction-tuning datasets, a subject the community is already extensively exploring. We focus on scaling from a **few thousand** to **millions** of samples that can be used for **pre-training LLMs from scratch**. This presents a unique set of challenges. 
```
Can add a link to this once the other post is published

File: cosmopedia.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+## Behind the scenes of Cosmopedia’s creation
+
+Besides the lack of information about the creation of the Phi datasets, another downside is that they use proprietary models to generate the data. To address these shortcomings, we introduce Cosmopedia, a dataset of synthetic textbooks, blog posts, stories, posts, and WikiHow articles generated by [Mixtral-8x7B-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1). It contains over 30 million files and 25 billion tokens, making it as the largest open synthetic dataset to date.
```
Comment (davanstrien) [inline cosmopedia.md:32]:
```suggestion
Besides the lack of information about the creation of the Phi datasets, another downside is that they use proprietary models to generate the data. To address these shortcomings, we introduce Cosmopedia, a dataset of synthetic textbooks, blog posts, stories, posts, and WikiHow articles generated by [Mixtral-8x7B-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1). It contains over 30 million files and 25 billion tokens, making it the largest open synthetic dataset to date.
```

File: cosmopedia.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+Heads up: If you are anticipating tales about deploying large-scale generation tasks across hundreds of H100 GPUs, in reality most of the time for Cosmopedia was spent on meticulous prompt engineering.
```
Comment (davanstrien) [inline cosmopedia.md:34]: <3

File: cosmopedia.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+### Prompts curation
+
+Generating synthetic data might seem straightforward, but maintaining diversity, which is crucial for optimal performance, becomes significantly challenging when scaling up. Therefore, it's essential to curate diverse prompts that cover a wide range of topics and minimize duplicate outputs, as we don’t want to spend compute on generating billions of textbooks only to discard most because they resemble each other closely. Before we launched the generation on hundreds of GPUs, we spent a lot of time iterating on the prompts with tools like HuggingChat. In this section we'll go over the process for creating over 30 million prompts for Cosmopedia, spanning hundreds of topics and achieving less than 1% duplicate content.
```
Comment (davanstrien) [inline cosmopedia.md:38]:
```suggestion
Generating synthetic data might seem straightforward, but maintaining diversity, which is crucial for optimal performance, becomes significantly challenging when scaling up. Therefore, it's essential to curate diverse prompts that cover a wide range of topics and minimize duplicate outputs, as we don’t want to spend compute on generating billions of textbooks only to discard most because they resemble each other closely. Before we launched the generation on hundreds of GPUs, we spent a lot of time iterating on the prompts with tools like [HuggingChat](https://huggingface.co/chat/). In this section, we'll go over the process of creating over 30 million prompts for Cosmopedia, spanning hundreds of topics and achieving less than 1% duplicate content.
```

File: cosmopedia.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+<p align="center">
+ <img src="https://huggingface.co/datasets/HuggingFaceTB/images/resolve/main/cosmopedia/text_clustering.png" alt="text-clustering" style="width: 60%; height: auto;"><br>
+ <em>Figure 9. The pipleline of text-clustering.</em>
+</p>
+
+#### Textbooks generation at scale
+
+We leverage [llm-swarm](https://github.com/huggingface/llm-swarm) library to generate 25 billion tokens of synthetic content using  [Mixtral-8x7B-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1). This is a tool for scalable synthetic data generation using local LLMs or inference endpoints on the Hugging Face Hub. It supports [TGI](https://github.com/huggingface/text-generation-inference) and [vLLM](https://github.com/vllm-project/vllm) inference libraries. We deployed Mixtral-8x7B locally on H100 GPUs from the Hugging Face Science cluster with TGI. The total compute time for generating Cosmopedia was over 10k GPU hours.
```
Comment (davanstrien) [inline cosmopedia.md:134]:
```suggestion
We leverage the [llm-swarm](https://github.com/huggingface/llm-swarm) library to generate 25 billion tokens of synthetic content using  [Mixtral-8x7B-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1). This is a scalable synthetic data generation tool using local LLMs or inference endpoints on the Hugging Face Hub. It supports [TGI](https://github.com/huggingface/text-generation-inference) and [vLLM](https://github.com/vllm-project/vllm) inference libraries. We deployed Mixtral-8x7B locally on H100 GPUs from the Hugging Face Science cluster with TGI. The total compute time for generating Cosmopedia was over 10k GPU hours.
```

File: cosmopedia.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+#### Benchmark decontamination
+
+Given that we generate synthetic data, there is a possibility of benchmark contamination within the seed samples or the model's training data. To address this, we implement a decontamination pipeline to ensure our dataset is free of any samples from the test benchmarks.
+
+Similar to Phi-1, we identify potentially contaminated samples using a 10-gram overlap. After retrieving the candidates,  we employ `difflib.SequenceMatcher` to compare the dataset sample against the benchmark sample. If the ratio of `len(matched_substrings)` to `len(benchmark_sample)` exceeds 0.5, we discard the sample. This decontamination process is applied across all benchmarks evaluated with the Cosmo-1B model, including MMLU, HellaSwag, PIQA, SIQA, Winogrande, OpenBookQA, ARC-Easy, and ARC-Challenge.
```
Comment (davanstrien) [inline cosmopedia.md:164]:
```suggestion
Similar to Phi-1, we identify potentially contaminated samples using a 10-gram overlap. After retrieving the candidates,  we employ [`difflib.SequenceMatcher`](https://docs.python.org/3/library/difflib.html) to compare the dataset sample against the benchmark sample. If the ratio of `len(matched_substrings)` to `len(benchmark_sample)` exceeds 0.5, we discard the sample. This decontamination process is applied across all benchmarks evaluated with the Cosmo-1B model, including MMLU, HellaSwag, PIQA, SIQA, Winogrande, OpenBookQA, ARC-Easy, and ARC-Challenge.
```

**Review by davanstrien — CHANGES_REQUESTED** [review]
Made some small suggestions (some are grammar nits you can ignore). Really like the post!

**Review by davanstrien — APPROVED** [review]
